synapse/handlers/saml_handler.py

```diff
@@ -290,17 +290,18 @@ async def saml_response_to_remapped_user_attributes(
                 if users:
                     registered_user_id = list(users.keys())[0]
                     logger.info("Grandfathering mapping to %s", registered_user_id)
-                    await self.store.record_user_external_id(
-                        self._auth_provider_id, remote_user_id, registered_user_id
-                    )
                     return registered_user_id
 
+            return None
+
+        with (await self._mapping_lock.queue(self._auth_provider_id)):
```

Comment:
I wonder if this lock should be moved into `SsoHandler`? (Aka should OIDC be using it also?)